### DISCUSSION TITLE
cant open file in subdirectory

### DISCUSSION THREAD

**Discussion by hirohitonumberonefan** [vercel/community#814] [Help] [answered]
so basically i have this line of code that opens a file in a subdirectory.
however apparently i get `[ERROR] OSError: [Errno 30] Read-only file system: 'database/995579769682935808/config.yml`
from
`self.config_file = open(f"{cwd}database/{Client.id}/config.yml", "r+")`
the program works without issues on my mac. what am i doing wrong here?
btw files that are in the same folder as the main file work normally
heres my vercel.json
`{
    "version": 2,
    "builds": [
        {
            "src": "./index.py",
            "use": "@vercel/python"
        }
    ],
    "routes": [
        {
            "src": "/",
            "dest": "/"
        }
    ]
}
`
(and yes, those files do exist)

**Comment by amyegan** [comment, marked as answer]
Hi @iylun! It looks like you might be trying to open a file that you can access from your local development environment, but not from a Severless environment. Examples: https://vercel.com/guides/how-can-i-use-files-in-serverless-functions

Are you able to share your repo or a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) so we can get a closer look at what's happening?

**Comment by ralpho79** [comment]
Facing the same issue - though to be specific - I cannot write to the file.
Here's my minimal reproducible example:
https://github.com/ralpho79/vercel-814-mre

And Vercel deploy:
https://vercel-814-mre.vercel.app/

I get the read-only error:

> [ERROR]	2023-02-08T16:04:01.425Z	55176f65-81de-4916-b825-21ed86630fcd	Exception on / [POST]
> Traceback (most recent call last):
>   File "/var/task/flask/app.py", line 2525, in wsgi_app
>     response = self.full_dispatch_request()
>   File "/var/task/flask/app.py", line 1822, in full_dispatch_request
>     rv = self.handle_user_exception(e)
>   File "/var/task/flask/app.py", line 1820, in full_dispatch_request
>     rv = self.dispatch_request()
>   File "/var/task/flask/app.py", line 1796, in dispatch_request
>     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
>   File "./api/index.py", line 11, in index
>     list_to_csv(text,"tmp/example.csv")
>   File "./api/index.py", line 19, in list_to_csv
>     with open(name, "w", newline="") as file:
> OSError: [Errno 30] Read-only file system: 'tmp/example.csv'

Seen somewhere this might have something to do with the fact that "On AWS Lambda the file system is always read-only except for the /tmp path."
[source](https://stackoverflow.com/questions/73394593/aws-lambda-function-returns-errormessage-errno-30-read-only-file-system)

But ... my 'tmp' does not seem to work either so not sure what else (config?) would I need to do.

**Reply by amyegan** [reply]
Thanks for including an MRE, @ralpho79! I got it to work for me by adding a `/` to the start of the tmp file path. So `list_to_csv(text,"tmp/example.csv")` subtly changed to `list_to_csv(text,"/tmp/example.csv")`

**Reply by ralpho79** [reply]
Amy thanks for checking!
I've extended the scope to verify the change and wanted to test both read & save.
Same repo.

The conclusion for me at the moment is - works on local vercel but not on public :(

My **local** vercel is using `tmp/example.csv` twice so I read and update the same file.

But when I deploy on **public** vercel, this no longer works
I can adjust the write function as per your comment but then when I use `list_to_csv(text,"/tmp/example.csv")` in one place and `csv_to_list("tmp/example.csv")` in another, I seem to end up with two different files (in two different places). 
Using `list_to_csv(text,"tmp/example.csv")` throws the Read-only file system error.

I found this on AWS which may explain what's going on..
https://repost.aws/questions/QUyYQzTTPnRY6_2w71qscojA/read-only-file-system-aws-lambda

Guess the best way could be to move to MongoDB or something instead of file storage.

**Reply by amyegan** [reply]
You need to use `/tmp` rather than `tmp` in all instances to access temporary storage in your serverless function. The guide on [using temporary storage in serverless functions](https://vercel.com/guides/how-can-i-use-files-in-serverless-functions#using-temporary-storage) has further explanation
